### ISSUE TITLE
Optimized Connection: pass over path instead of Workbook object

### ISSUE THREAD

**Issue by fzumstein** [ZoomerAnalytics/xlwings#277]
This should probably get rid of zombie processes as described in #236 or when the Workbook() is created as modulewide global outside of a function.

**Comment by sdementen** [comment]
Could you elaborate a bit on the issue and how it would be solved by not passing the workbook ? Tx

**Comment by fzumstein** [comment]
It's not about *passing* a workbook object but rather having it as a global sitting in `_xlwindows.py`. For whatever reason Python doesn't manage to properly garbage collect it which leaves a COM Reference to Excel which in turn prevents the Excel process from shutting down when it should. I don't understand the details enough, but in my opinion it's worth to try out the option of saving a simple string instead of a COM object as it should be an easy change.

**Comment by sdementen** [comment]
Ok, it is the Py.SetAttr thing in the vba code that make it stick in memory !
Could this be solved alternatively by removing the reference as soon as caller() is used ? Or adding the workbook to a global list object from which the workbook is popped from within python (again, the objective being to remove the reference after being passed) ?
I have the feeling (no hard facts) that passing directly the workbook is simpler than going through the more complex path+hwnd+accessiblefromwindow machinery.

**Comment by aafwb** [comment]
I am not yet a proficient Pythonista versed in the core language and its
libraries, but I have seen some video tutorials on weakref which elaborate
on issues relating with caching and gc and wondered whether this could be
worthy of further investigation?

Apologies if this comment is not relevant - just curious and trying to be
helpful.
Kind regards,
Alberto

**Comment by ericremoreynolds** [comment]
Hi guys,

in detail the problem is the following:

(a) the Python process gets killed when the Excel process terminates

(b) the Excel process terminates only when two conditions are met (i) the user has closed all the windows (ii) there are no remaining COM references to any object related to that Excel process.

This clearly creates a problem, because in order to control Excel, the Python process must have a COM reference to it. If it stores this in a global variable, this COM reference remains alive until the Python process is shut down, but the Python process is not shut down as long as the the COM reference remains alive. So you have a cycle.

There are various ways of solving this I guess, for example, after the call the `caller` could be unset, or the workbook could be reachable not through a COM reference but through a path or something like that. We are still in the process of deciding which approach is best. I agree with @sdementen that a COM Workbook reference feels like the simplest solution, but it remains to be seen how workable it really is.

@aafwb weakrefs would not help unfortunately, it's a COM reference counting cycle (sort of) and doesn't have much to do with the Python GC.

**Comment by sdementen** [comment]
Would a
Py.SetAttr Py.Module("xlwings._xlwindows"), "xl_workbook_current", False
after the Py.Exec fit the bill ?

**Comment by ericremoreynolds** [comment]
Yeah, I'm not sure it's perfect (e.g. the user could still save another reference in a global anyway) but it's and OK temporary solution.

**Comment by fzumstein** [comment]
Neither approach seems to work for me for a script like this:

```
from xlwings import Workbook, Range
wb = Workbook.caller()
def rand_numbers():
    Range('A1').value = 1
```

when I set `global wb` and `wb = None` from inside the function, Excel even crashes in my case. Keeping Workbook.caller() inside the function seems like the most reliable approach for right now.

**Comment by sdementen** [comment]
I noticed also rather often that I have still an Excel zombie when using the standard approach (not the optimized connection): I call RunPython that starts a python process that connects back through excel via the Workbook.caller(), does some work and then ends. When I then close Excel, it stays as zombie.
Isn't there another issue then ?

Otherwise, with optimized connection, we could call Py.Quit (ericremoreynolds/excelpython#59) before closing Excel to first kill the python process and then Excel would quit gracefully.

(ps: i am not always sure of what I am talking about here ;-)

**Comment by fzumstein** [comment]
I like the idea with `Py.Quit` - @ericremoreynolds, what do you think?

As with regards to the zombie process without using the COM server: I am happy to look into it, if you can replicate it reliably. But as far as I am aware, the only time that should happen is if the Python interpreter is being killed forcibly (see #55).
Other than that, if you work on unit tests (that run with the application in hidden mode) and don't properly close a Workbook, this can also leave you with the impression of leaving a zombie process.

**Comment by sdementen** [comment]
I'll try to reproduce the zombie behavior on a simple example (but my app
is rather complex). However, it happens on a basic flow : excel calls
python, python runs, python exits , excel is then closed manually but excel
stays zombie.
Issue #55 describes an issue where excel is stopped before python, correct
? in this case, I guess the app.Quit is equivalent to an app.Visible=False
as the process is still there ?

But could my problem depend on the way python exits (gracefully or via a
sys.exit call) ? anyway, better to try to reproduce the bug on a simple
case.

**Comment by fzumstein** [comment]
yup, the way that python exits sounds exactly like the perfect reason why you see that issue...
